### DISCUSSION TITLE
Cache bounding boxes for "senses"

### DISCUSSION THREAD

**Discussion by akhilman** [emilk/egui#724] [Ideas]
We can save the bounding boxes of widgets during rendering in lists (or maybe a binary tree) for each sense type. So in the next frame, we could cheaply check if the GUI will respond to user input and skip all rendering if not needed.

**Comment by emilk** [comment]
This is a pretty neat idea for saving some CPU in the "reactive" mode for when the egui is not covering the entire screen (e.g. in a game).

**Reply by Raj2032** [reply]
Has this feature been implemented yet?

**Comment by akhilman** [comment]
I think It could work also with a full screen UI. We may skip the mouse events when the cursor travels over the widgets that have not sense for "mouse over".

**Reply by akhilman** [reply]
I am interested in implementing it in the future

**Reply by emilk** [reply]
Currently all widgets sense mouse over (hover). This is so that one can attach a tooltip on them after adding them: `ui.add(widget).on_hover_text(…)`.

We could split it into two things: sensing hovering (all widgets) and sensing hover positions (e.g. Plot). We would still need to repaint when the mouse goes from one sensing area to another (because we may need to show a different tooltip).

**Reply by mankinskin** [reply]
Can we really assume that the widgets do not change when there is no sense event? The user might have changed shared data which requires the widget to be redrawn, even when it wasn't interacted with.

Maybe for this we would need something like a "force redraw" api, the user has to call explicitly when he updates shared data. Possibly something like an observer pattern would be useful too.

**Reply by akhilman** [reply]
If the user has changed something, it will trigger a redraw. With senses and boundaries, we can filter input events such as mouse movement when the application is not doing anything.

**Reply by mankinskin** [reply]
We do already have reactive mode, and actually also "request_redraw" to force a redraw.

I don't fully understand the idea yet.. do you want to redraw only widgets that have been interacted with this frame? Or something different?

**Reply by akhilman** [reply]
The idea is to reduce the number of redraws in reactive mode.

Right now, in reactive mode, the egui redraws window on every event. For example, when the mouse cursor moves over the window, each frame is redrawn, regardless of whether there will be any reaction to the mouse move event. I suggest using bounding boxes and event sensing flags to filter events so that redrawing occurs only when the widget actually responds to the event. So if the widget under the cursor senses the "mouse hover" event, then we need only two events, when the mouse enters the widget's borders and when it leaves them. 

Redrawing only changed widgets will require caching and invalidating the cache, I'm not sure if this is easily achievable in immediate mode.

**Reply by mankinskin** [reply]
Got it.. sounds like a good idea. Asynchronous calls could just call `request_repaint` to show their changes.

Yeah, something like a cache for the last frame would definitely be helpful for a lot of things. And it shouldn't be too hard to do, I'd think.. basically just store all of the Responses after each frame?

**Reply by Raj2032** [reply]
@akhilman Are you planning to implement such a feature or do I have to use "bounding boxes"?

**Reply by akhilman** [reply]
> @akhilman Are you planning to implement such a feature or do I have to use "bounding boxes"?

Not yet. I'm waiting for more details on #843 that might help me.
If you want to implement this feature yourself, you are welcome.

**Reply by Raj2032** [reply]
@akhilman Truth be told I never used this GUI library at all, I am considering using it. I might use it for a project (I can then become more familiar with it) and then I can help out. 

I was looking at `lru-rs` https://crates.io/crates/lru which caches things, I don't know if this would help or not.

**Comment by akhilman** [comment]
Implementation could be shared with #1250.

**Reply by Raj2032** [reply]
Is this similar to retained mode?

**Reply by akhilman** [reply]
Not really. We are still drawing everything every frame, but this cache will help us to skip some frames triggered by mouse events.

**Reply by Raj2032** [reply]
AH right mate
